### DISCUSSION TITLE
Plan to implement a reset controller in Zephyr

### DISCUSSION THREAD

**Discussion by andrei-edward-popa** [zephyrproject-rtos/zephyr#42644] [Ideas]
Is there any plan to introduce a reset controller API driver in Zephyr? I have this in mind and I want to try to implement such a driver, but I don't know if it is some work in progress right now.

**Comment by henrikbrixandersen** [comment]
By reset controller do you mean resetting the SoC on which Zephyr is running? If so, we already have `sys_reboot()` which call architecture-dependent functions for disabling IRQs, clocks and resetting the SoC. See https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/reboot.c

**Reply by andrei-edward-popa** [reply]
No, by reset controller I mean a driver that resets other device drivers. Let me try to create an example.

    reset: reset-controller@00000000 {
        compatible = "reset";
        reg = <0x00000000>;
        ...
    }

    uart0: uart@00000001 {
        ...
        resets = <&reset REG_RESET_OFFSET REG_RESET_VALUE>;
        ....
    }

Now when you want to use the reset controller driver, you will need to get in the uart driver implementation the device with DT_DRV_COMPAT = reset to use the reset controller driver API to reset the uart device. The 00000000 need to be the base address of the SoC's reset registers.

**Reply by carlocaione** [reply]
yes, this is something missing in Zephyr. I have seen devs using syscon because we are missing a proper API for reset controllers. You have my +1 if you want to work on this. Please add me as reviewer in that case.

**Reply by andrei-edward-popa** [reply]
Cool, I will start next week.

**Reply by andrei-edward-popa** [reply]
@carlocaione Hi! Here is PR for first implementation of the reset controller: https://github.com/zephyrproject-rtos/zephyr/pull/43073
